src/javaharness/java/arcs/android/impl/LocalArcsImpl.java

```diff
@@ -0,0 +1,53 @@
+package arcs.android.impl;
+
+import javax.inject.Inject;
+
+import arcs.api.Arc;
+import arcs.api.ArcsEnvironment;
+import arcs.api.ArcsImpl;
+import arcs.api.PECInnerPort;
+import arcs.api.PecPortManager;
+import arcs.api.PortableJson;
+
+public class LocalArcsImpl extends ArcsImpl {
+
+  @Inject protected ArcsEnvironment environment;
+  @Inject protected PecPortManager pecPortManager;
+
+  @Inject LocalArcsImpl() {}
+
+  @Override
+  public void runArc(Arc arc) {
+    PortableJson request =
+        jsonParser
+            .emptyObject()
+            .put("message", "runArc")
+            .put("arcId", arc.getArcId())
+            .put("pecId", arc.getPecId())
+            .put("recipe", arc.getRecipe());
+
+    if (arc.getParticleName() != null && arc.getParticleId() != null) {
```

Comment:
Is it an error for one to be provided but not the other? Should probably add an error message in that case.

Reply:
this is handled by the builder - if only particleName was set, particleId will be generated.